cpp/src/Ice/LoggerMiddleware.cpp

```diff
@@ -32,85 +40,109 @@ LoggerMiddleware::dispatch(Ice::IncomingRequest& request, function<void(Outgoing
                 }
                 sendResponse(std::move(response));
             });
     }
     catch (const UserException&)
     {
-        // No warning.
+        if (_traceLevel > 0)
+        {
+            logDispatch(ReplyStatus::UserException, request.current());
+        }
         throw;
     }
     catch (const RequestFailedException& ex)
     {
-        if (_warningLevel > 1)
+        if (_traceLevel > 0 || _warningLevel > 1)
         {
-            warning(ex, request.current());
+            logDispatchException(ex, request.current());
         }
         throw;
     }
-    catch (const Ice::Exception& ex)
+    catch (const Ice::LocalException& ex)
     {
-        warning(ex, request.current());
+        logDispatchException(ex, request.current());
         throw;
     }
     catch (const std::exception& ex)
     {
-        warning(ex.what(), request.current());
+        logDispatchException(ex.what(), request.current());
         throw;
     }
     catch (...)
     {
-        warning("c++ exception", request.current());
+        logDispatchException("c++ exception", request.current());
         throw;
     }
 }
 
 void
-LoggerMiddleware::warning(const Exception& ex, const Current& current) const noexcept
+LoggerMiddleware::logDispatch(ReplyStatus replyStatus, const Current& current) const noexcept
+{
+    Trace out{_logger, _traceCat};
+    out << "dispatch of " << current.operation << " to ";
+    printTarget(out, current);
+    out << " returned a response status with reply status " << replyStatus;
+}
+
+void
+LoggerMiddleware::logDispatchException(string_view exceptionDetails, const Current& current) const noexcept
 {
-    Warning out(_logger);
-    out << "dispatch exception: " << ex;
-    warning(out, current);
+    Warning out{_logger};
+    out << "failed to dispatch " << current.operation << " to ";
+    printTarget(out, current);
+
+    if (!exceptionDetails.empty())
+    {
+        out << ":\n" << exceptionDetails;
+    }
 }
 
 void
-LoggerMiddleware::warning(const string& exceptionDetails, const Current& current) const noexcept
+LoggerMiddleware::logDispatchException(const LocalException& ex, const Current& current) const noexcept
 {
-    Warning out(_logger);
-    out << "dispatch exception: " << exceptionDetails;
-    warning(out, current);
+    ostringstream os;
+    os << ex;
+    logDispatchException(os.str(), current);
 }
 
 void
-LoggerMiddleware::warning(Warning& out, const Current& current) const noexcept
+LoggerMiddleware::printTarget(LoggerOutputBase& out, const Current& current) const noexcept
 {
-    out << "\nidentity: " << identityToString(current.id, _toStringMode);
-    out << "\nfacet: " << escapeString(current.facet, "", _toStringMode);
-    out << "\noperation: " << current.operation;
+    out << identityToString(current.id, _toStringMode);
+
+    if (!current.facet.empty())
+    {
+        out << " -f " << escapeString(current.facet, "", _toStringMode);
+    }
 
     if (current.con)
```

Comment:
Should we print `over collocated` or similar, when there is no connection?

Reply:
We could also print over `<transport name>` when the transport is not IP-based.